### PR TITLE
Add MSBuild properties for common ILLink arguments

### LINKED ISSUE #11006: Add msbuild properties for common illink arguments
Update illink task to have a set of known properties for common arguments. This is to avoid putting everything into ExtraArgs and it's easier to tweak by the developers or 3rd party projects.

Map to existing properties should be prefered where it makes sense (e.g. verbose logging).

For inspiration csc task properties - https://docs.microsoft.com/en-us/visualstudio/msbuild/csc-task?view=vs-2019#parameters

@sbomer

### PR DESCRIPTION
**Pull request by sbomer**
For now this just adds private properties to make it easier for SDK components to set ILLink arguments (instead of using `_ExtraTrimmerArgs`). Most of them are not set by default, with the exception of `_TrimmerDefaultAction` which replaces the `-c` and `-u` flags.

I don't think there is much to test here, since most of the new properties are not tied to existing SDK logic. An existing test checks the behavior for `_TrimmerDefaultAction`, and I have added a sanity check that the boolean options for the optimizations are actually used by validating that they produce errors if they're not boolean. The functionality of the options themselves is tested in https://github.com/mono/linker.

fixes https://github.com/dotnet/sdk/issues/11006.

### CODE REVIEW

**Comment by dsplaisted** [comment]
Is there a PR for ILLink that you can link to where these options were added to the task?

**Comment by sbomer** [comment]
@dsplaisted here it is: https://github.com/mono/linker/pull/1057

**Review by marek-safar — APPROVED** [review]

**Comment by dsplaisted** [comment]
Does this complete #11006?

**Comment by dsplaisted** [comment]
Looks like some tests are failing, probably because they target .NET Core 3.x and are referencing a package which now requires .NET 5.  The tests may need to be retargeted, or possibly split out so we are testing both .NET Core 3.x and .NET 5 scenarios.

**Comment by sbomer** [comment]
Yes, this fixes https://github.com/dotnet/sdk/issues/11006. The test failures had me pretty confused for a bit - I think the issue is that the referenced project is shared across all of those tests, which wasn't the intention. When I created it for 5.0 in the new test, some of the existing testcases were picking that up (probably when running in parallel). I'm going to fix it so that there's a separate reference project for each testcase.

**Review by swaroop-sridhar — APPROVED** [review]

**Comment by sbomer** [comment]
@dsplaisted PTAnotherL

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs [hunk 1]
```diff
@@ -441,8 +453,35 @@ private void EnableNonFrameworkTrimming(XDocument project)
                                                  new XElement("action"))));
         }
 
-        private TestProject CreateTestProjectForILLinkTesting(string targetFramework, string mainProjectName, string referenceProjectName, bool usePackageReference = true)
+        private TestProject CreateTestProjectForILLinkTesting(
+            string targetFramework,
+            string mainProjectName,
+            string referenceProjectName = null,
+            bool usePackageReference = true,
+            [CallerMemberName] string callingMethod = "")
```
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:461]: If you are calling this method from a `Theory`, then the different instances will still collide on the path.  The pattern we generally use for this is to also pass in an optional `identifier` parameter, which can be a concatenation of the `Theory` parameters.

**Review by dsplaisted — APPROVED** [review]

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs (hunk 1 above)
Comment (sbomer) [reply src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:461]: Thanks!

**Pull request merged**